server.js

```diff
@@ -24,12 +19,143 @@ const app = express();
+    isbn: Number,
+    isbn13: Number,
+    language_code: String,
+    num_pages: Number,
+    ratings_count: Number,
+    text_reviews_count: Number
+})
+
+
+// Function to start seeding our Database, will only run if
+// RESET_DB environment variable is present and is true
+if(process.env.RESET_DB){
+  // const seedDatabase = async () => {
+  //   await Book.deleteMany({})
+    booksData.forEach(item => {
+      const newBook = new Book(item)
+      newBook.save()
+    })
+  }
+  // seedDatabase()
+// }
+
+
+
+//second endpoint to get all books
+app.get('/books', async (req,res) => {
+  const booksToShow = await Book.find({})
+  res.json(booksToShow)
+})
+
+//get one book based on id
+app.get('/books/id/:id', async (req, res) => {
+try{
+  const bookById =  await Book.findById(req.params.id)
+  if(bookById) {
+    res.status(200).json({response: bookById, success:true})
+  } else {
+    res.status(404).json({response: 'Book not found!', success:false})
+  }
+} catch(err) {
+  res.status(400).json({response: 'Error!', success:false})
+}
+
+})
+
+//get the books based on rating
+app.get('/books/rating', async (req, res) => {
+  let booksToShow = await Book.find(req.query)
+
+  //books with rating greater than given value
+  if (req.query.average_rating) {
+    const booksByRating= await Book.find().gt(
+      'average_rating',
+      req.query.average_rating
+    )
+    booksToShow = booksByRating
+  }
+
+  res.json(booksToShow)
+})
+
 // Start defining your routes here
-app.get("/", (req, res) => {
-  res.send("Hello Technigo!");
+  app.get("/", (req, res) => {
+    res.json({
+      message: 'This is API for book data endpoints',
+      data: listEndpoints(app)
+    })
 });
 
+
+
+// //get data by book's name and author using query parameters
+app.get("/books", (req, res) => {
+  const {bookName, bookAuthor} = req.query
+  let booksToSend = booksData
+
+  if(bookName) {
+    booksToSend = booksData.filter((item) => item.title.toLowerCase().indexOf(bookName.toLowerCase()) !== -1)
+  }
+
+  if(bookAuthor) {
+    booksToSend = booksToSend.filter((item) => item.authors.toLowerCase().indexOf(bookAuthor.toLowerCase()) !== -1)
+  }
+
+  res.json({
+    response: booksToSend,
+    success:true
+  })
+  
+});
+
+// //get book data by id
+app.get('/books/id/:bookID', (req, res) => {
+  const { bookID } = req.params;
+  console.log(req)
+  const bookById = booksData.find((item) => item.bookID === +bookID);
+
+  if (!bookById) {
+    res.status(404).send('No book found!')
+  };
+  
+  res.json(bookById);
+})
+
+//get bbok data by name
+app.get('/books/name/:bookName', (req, res) => {
+  const { bookName } = req.params
+  const bookByName = booksData.find((item) => item.title.toLowerCase() === bookName.toLowerCase())
+
+  if(!bookName) {
+    res.status(404).json({
+      response: 'No book found with that name',
+      success: false
+    }) 
+  } else {
+    res.status(200).json({
+      response: bookByName,
+      success: true
+    }) 
+  }
```

Comment:
I was unfortunately not able to fetch a specific book with the title either. I notice that you're mapping through the json data here too. You seem to have the data in the database because returning all the books with the main query and fetching the id with the mongoose id works, so perhaps there's something not working because of that? I think you could use the mongoose query methods for fetching this data from the database?

You could do this by fetching the book by name from the database using regex, something like this for example:

`const { bookName } = req.params
const bookByName = await Book.find({
title: new Regexp(bookName, "i")
})`

Or you could change your const on row 62 to a let and then do if statements inside your main query and then assign your let booksToShow to be your result of the query according to the if statements. Something like this:

`let booksToShow = await Book.find(req.query
if (req.query.bookName) {
const bookByName = await Book.find({
title: new Regexp(req.query.bookName, "i")
})
booksToShow = bookByName`

It also gave me a "successful" request as the message from that came back was success: true even though I was not able to fetch a book. So perhaps also wrapping the whole request into a try catch form would help in spotting what went wrong with the request for the user too?  😊